ambry-server/src/integration-test/java/com/github/ambry/server/ServerTestUtil.java

```diff
@@ -2051,6 +2059,290 @@ static void endToEndReplicationWithMultiNodeSinglePartitionTest(String routerDat
+      PartitionId partitionId = blobId1.getPartition();
+      controlReplicationForPartition(channels, partitionId, false);
+
+      // Now send the delete to two data nodes in the same DC
+      List<DataNodeId> toBeDeleteDataNodes = dataNodesPerDC.values().stream().findFirst().get();
+      Pair<SSLConfig, SSLSocketFactory> pair = sslSettingPerDC.get(toBeDeleteDataNodes.get(0).getDatacenterName());
+      ConnectedChannel channel1 =
+          getBlockingChannelBasedOnPortType(portType, toBeDeleteDataNodes.get(0), pair.getSecond(), pair.getFirst());
+      channel1.connect();
+      ConnectedChannel channel2 =
+          getBlockingChannelBasedOnPortType(portType, toBeDeleteDataNodes.get(1), pair.getSecond(), pair.getFirst());
+      channel2.connect();
+      DeleteRequest deleteRequest1 = new DeleteRequest(1, "deleteClient", blobId1, System.currentTimeMillis());
+      channel1.send(deleteRequest1);
+      DeleteResponse deleteResponse = DeleteResponse.readFrom(new DataInputStream(channel1.receive().getInputStream()));
+      assertEquals(ServerErrorCode.No_Error, deleteResponse.getError());
+      DeleteRequest deleteRequest2 =
+          new DeleteRequest(1, "deleteClient", blobId1, deleteRequest1.getDeletionTimeInMs());
+      channel2.send(deleteRequest2);
+      deleteResponse = DeleteResponse.readFrom(new DataInputStream(channel2.receive().getInputStream()));
+      assertEquals(ServerErrorCode.No_Error, deleteResponse.getError());
+
+      // Now send the undelete operation through router, and it should fail because of not deleted error.
+      Future<Void> future = router.undeleteBlob(blobId1.toString(), "service", null);
+      try {
+        future.get();
+        fail("Undelete blob " + blobId1.toString() + " should fail");
+      } catch (ExecutionException e) {
+        assertTrue(e.getCause() instanceof RouterException);
+        assertEquals(RouterErrorCode.BlobNotDeleted, ((RouterException) e.getCause()).getErrorCode());
+      }
+
+      // Now see if either data node 1 or data node 2 has undelete or not, if so, undelete would replicate. If not,
+      // delete would replicate.
+      List<PartitionRequestInfo> partitionRequestInfoList = getPartitionRequestInfoListFromBlobId(blobId1);
+      boolean hasUndelete = false;
+      for (ConnectedChannel connectedChannel : new ConnectedChannel[]{channel1, channel2}) {
+        GetRequest getRequest =
+            new GetRequest(1, "clientId1", MessageFormatFlags.BlobProperties, partitionRequestInfoList,
+                GetOption.Include_All);
+        channel1.send(getRequest);
+        GetResponse getResponse =
+            GetResponse.readFrom(new DataInputStream(channel1.receive().getInputStream()), clusterMap);
+        assertEquals(ServerErrorCode.No_Error, getResponse.getPartitionResponseInfoList().get(0).getErrorCode());
+        try {
+          BlobProperties propertyOutput = MessageFormatRecord.deserializeBlobProperties(getResponse.getInputStream());
```

Comment:
Does this only test deserialization of blob properties? It seems `propertyOutput` is never used.